command_line/generate_mask.py

```diff
@@ -32,108 +30,157 @@
-        if len(imagesets) > 1:
-            # Check beams (for resolution) and detectors are equivalent in each case
-            # otherwise the mask may not be appropriate across all imagesets
-            detectors = experiments.detectors()
-            beams = experiments.beams()
-            for d in detectors[1:]:
-                if not d.is_similar_to(detectors[0]):
-                    raise Sorry(
-                        "multiple imagesets are present, but their detector"
-                        " models differ"
-                    )
-            for b in beams[1:]:
-                if not b.is_similar_to(beams[0]):
-                    raise Sorry(
-                        "multiple imagesets are present, but their beam"
-                        " models differ"
-                    )
-
-        imageset = imagesets[0]
-
-        # Generate the mask
-        generator = MaskGenerator(params)
-        mask = generator.generate(imageset)
-
-        # Save the mask to file
-        print("Writing mask to %s" % params.output.mask)
-        with open(params.output.mask, "wb") as fh:
-            pickle.dump(mask, fh)
-
-        # Save the experiment list
-        if params.output.experiments is not None:
-            for imageset in imagesets:
-                imageset.external_lookup.mask.data = ImageBool(mask)
-                imageset.external_lookup.mask.filename = params.output.mask
-            from dxtbx.model.experiment_list import ExperimentListDumper
-
-            print("Saving experiments to {0}".format(params.output.experiments))
-            dump = ExperimentListDumper(experiments)
-            dump.as_file(params.output.experiments)
+def script(experiments, params):
+    """
+    Generate a pixel mask for each image in an experiment.
+
+    Use the masking parameters :param:`params` and an experiment in the experiment
+    list :param:`experiments` to define pixel masks for the associated imagesets.
+    The masks are generated using :mod:`dials.util.masking`.
+
+    The masks will be saved to disk at the location specified by
+    :attr:`params.output.mask`.  Optionally, if a path
+    :attr:`params.output.experiments` is set, a modified copy of :param:`experiments`
+    with the masks applied will be saved to that location.
+
+    Args:
+        experiments (:type:`dxtbx_model_ext.ExperimentList`): An experiment list
+            containing only one :type:`dxtbx_model_ext.Experiment` object.
+        params (:type:`libtbx.phil.scope_extract`): Masking parameters, having the
+            structure defined in :data:`dials.util.masking.phil_scope`.
+
+    Returns:
+        Tuple[:rtype:`scitbx_array_family_flex_ext.bool`]: A tuple containing the
+            generated pixel masks.
+        :rtype:`dxtbx_model_ext.ExperimentList` (optional): A copy of
+            :param:`experiments` with the masks applied.  Only returned if
+            :attr:`params.output.experiments` is set.
```

Comment:
[**tl;dr** - type is probably wrong for return mask, also return type in general]

Couple of Docstring notes; not really criticism as points not raised before:
- I don't think you need to explicitly link the types as types in the type field. At least, it should know what these are types already? Even if it doesn't, it should just be a matter of parser intelligence. Made redundant by,
- Probably a bigger discussion but I've started moving even more to Python3 style type annotations. The main checking tool is [mypy](http://mypy-lang.org/), which has a standard for [python2](https://mypy.readthedocs.io/en/stable/cheat_sheet.html) _**But** also_ a way to automatically convert the python2-style annotations to real [PEP-484](https://www.python.org/dev/peps/pep-0484/) annotations once the codebase is ready for python3-only, so it's a bit more future-compatible.
- On this note, there are definitely plugins for sphinx/napoleon that will read the type annotations and extract them automatically so that you don't need to define them in the actual docstring text. This means that they only ever need to be written once, and in a static-analysis-verifiable way.

Specific to this code point:
- Assuming that this really does return a tuple of multiple masks rather than a tuple of a single mask, you probably mean `Tuple[scitbx.array_family.flex.bool, ...]` (and actual import path should be fine)
- The return of a type that could be multiple things is
  - In sphinx notation `A or B`
  - In PEP-484 notation: `Union[A, B]`

So, the return type is:
```
Tuple[scitbx.array_family.flex.bool,...] or Tuple[Tuple[scitbx.array_family.flex.bool,...], dxtbx.model.ExperimentList]
or
Union[Tuple[Tuple[scitbx.array_family.flex.bool,...], dxtbx.model.ExperimentList], Tuple[scitbx.array_family.flex.bool,...]]
```
And the full type-annotation type for the function, mypy2-style is:
```python
# type: (dxtbx.model.ExperimentList, libtbx.phil.scope_extract) -> Union[Tuple[Tuple[scitbx.array_family.flex.bool,...], dxtbx.model.ExperimentList], Tuple[scitbx.array_family.flex.bool,...]]
```
or

```python
# mypy python 2-style type annotation:
def generate_mask(
    experiments,  # type: dxtbx.model.ExperimentList
    params  # type: libtbx.phil.scope_extract
):
    # type: (...) -> Union[Tuple[Tuple[scitbx.array_family.flex.bool,...], dxtbx.model.ExperimentList], Tuple[scitbx.array_family.flex.bool,...]]
    """
    Generate a pixel mask for each image in an experiment.

    ...
    """
```
which becomes, in python 3:
```python
def generate_mask(
    experiments: dxtbx.model.ExperimentList, params: libtbx.phil.scope_extract
) -> Union[
    Tuple[Tuple[scitbx.array_family.flex.bool, ...], dxtbx.model.ExperimentList],
    Tuple[scitbx.array_family.flex.bool, ...],
]:
    """
    Generate a pixel mask for each image in an experiment.

    ...
    """
```

This looks... not good, but can generally be simplified by scoped imports and custom return types e.g.
```python
from dxtbx.model import ExperimentList
Masks = Tuple[scitbx.array_family.flex.bool, ...]

def generate_mask(
    experiments: ExperimentList, params: libtbx.phil.scope_extract
) -> Union[Masks, Tuple[Masks, ExperimentList]]:
    """Docstring"""
```

This can obviously be reduced further if the return type is simplified to something like always returning a tuple (which is easier to use also) e.g.`Tuple[Masks, Optional[ExperimentList]]`

Reply:
The other point about not having lots of reST markup in the type field is that google/numpy style docstrings are specifically designed to be easy to read inline in code, which obviously the markup somewhat obfuscates

Reply:
Thanks, I got a bit confused when reading the various different specifications, so that's very helpful.  I'll go through again and strip out some chaff.

The returned object is indeed, in general, a tuple of multiple mask objects.  As to always returning a tuple with an optional second value, you have a good point.  I think I tried to over-engineer it a bit because I once got caught out by https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.polyfit.html#numpy-polyfit

Reply:
Ah, maybe the numpy style that explicitly means multiple/tuple? From the linked document I can't tell if it's one or the other or a tuple of both or what.

The typing module (in 3, but backported) makes stuff nicer, because all these type annotations may look slightly verbose but are completely legal python expressions so statically verifiable, which is why you can do stuff like `Masks = Tuple[flex.bool,...]` and have it work.

Reply:
But as an ambition, the actual markup the typing should make it unambiguous, and then whatever documentation generator can render it how it likes.

Reply:
Hmm... do we want to specify `IO[str]` or similar for functions that write to disk?

Reply:
As in https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html#miscellaneous

Reply:
Not sure if I've misunderstood the question, but that's still for taking parameters/returning values - just if it's a file-like object (e.g. something that has .write, .read, .seek, .flush maybe etc). So, for example, the `show` methods as discussed in https://github.com/cctbx/cctbx_project/issues/275 often take a `log` parameter that it writes to, and they would be of type `IO`. The PEP seems to specify it for *just* things that write to disk but people seem to use it more generally.

Now, if you want to actually encode the behaviour of a function's side effects into the type system, then you start getting into ML style languages and monads and what have you......

Reply:
Ah, yep.  I skim-read and didn't correctly comprehend.  Ignore the question!